extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -59,34 +70,73 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
                     public CompletionStage<Boolean> apply(Permission permission) {
                         return securityIdentity.checkPermission(permission);
                     }
-                })
-                .build();
+                }).build();
     }
 
     public CompletionStage<SecurityIdentity> authenticate(RoutingContext context,
             IdentityProviderManager identityProviderManager,
             DefaultTenantConfigResolver resolver) {
         Cookie sessionCookie = context.request().getCookie(SESSION_COOKIE_NAME);
+        TenantConfigContext configContext = resolver.resolve(context, true);
 
         // if session already established, try to re-authenticate
         if (sessionCookie != null) {
             String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
-            return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
-                    .thenCompose(new Function<SecurityIdentity, CompletionStage<SecurityIdentity>>() {
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
+            return authenticate(identityProviderManager, new IdTokenCredential(idToken, context))
+                    .handle(new BiFunction<SecurityIdentity, Throwable, SecurityIdentity>() {
                         @Override
-                        public CompletionStage<SecurityIdentity> apply(SecurityIdentity securityIdentity) {
-                            return CompletableFuture
-                                    .completedFuture(augmentIdentity(securityIdentity, tokens[1], tokens[2], context));
+                        public SecurityIdentity apply(SecurityIdentity identity, Throwable throwable) {
+                            if (isLogout(context, configContext)) {
+                                // force logout, the challenge will be built accordingly and redirect to the logout endpoint at the OP
+                                throw new AuthenticationFailedException();
+                            }
+
+                            // if authentication failed, the reason could be the token no longer being valid, so we try a 
+                            // silent refresh if required
+                            if (throwable != null) {
+                                if ((identity = trySilentRefresh(configContext, idToken, refreshToken, context)) == null) {
+                                    // if the refresh fails, we just propagate the original exception
+                                    if (throwable instanceof AuthenticationFailedException) {
```

Comment:
The code around here throw `AuthenticationCompletionException` to prevent the redirect loops, so that the user gets 401, otherwise a challenge will be created again and the OIDC will redirect again given that a user has just authenticated, we've had these problems before :-)

Reply:
Not sure, I think I'm covering this scenario in the test?

Reply:
@pedroigor but your test does not have a refresh failed scenario ? Try to update this code locally to assume it is a failure and see what happens

Reply:
It does ... The last checks if a redirect is performed due to the session being invalid at the OP.

Reply:
I mean a refresh grant returns an error, that what was causing a lot of headache earlier with the code grant returning an error

Reply:
That is what that last part does, the refresh will fail due to an invalid session (but can be anything) and the user properly redirected to the login page.

If the check is not enough, I'm happy to provide something else.

Reply:
@pedroigor I need to think more about it, what we definitely want to avoid is that the challenge is sent to Keycloak while the user session is still valid

Reply:
That is a valid use case ... Suppose the OP is down, local session active. When refreshing we are going to invalidate the session. A similar behavior than if using token introspection with bearer tokens.

Depending on the app requirements the behavior is expected as you must enforce that local sessions have their corresponding ones at the OP. Note that if you don't want this, you can just disable the functionality and fallback to local validation (as it stands today).

Reply:
Hi Pedro, @pedroigor thanks, sure that makes sense, what I'm somewhat concerned about is that redirection loop. I think if we are 100% sure that the refresh is attempted only when the ID token has expired then `AuthenticationFailedException` will get a user facing a KC/etc login page. This is one reason why we should check the exception message before a refresh.